javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/ExtensionClassLoader.java

```diff
@@ -29,33 +38,83 @@
   // configured, and so using slf4j here would initialize slf4j-simple before we have a chance to
   // configure the logging levels
 
-  public static ClassLoader getInstance(ClassLoader parent) {
+  static {
+    ClassLoader.registerAsParallelCapable();
+  }
+
+  public static ClassLoader getInstance(ClassLoader parent, File javaagentFile) {
+    List<URL> extensions = new ArrayList<>();
+
+    includeEmbeddedExtensionsIfFound(parent, extensions, javaagentFile);
+
     // TODO add support for old deprecated property otel.javaagent.experimental.exporter.jar
-    List<URL> extension =
+    extensions.addAll(
         parseLocation(
             System.getProperty(
                 "otel.javaagent.experimental.extensions",
-                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_EXTENSIONS")));
+                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_EXTENSIONS")),
+            javaagentFile));
 
-    extension.addAll(
+    extensions.addAll(
         parseLocation(
             System.getProperty(
                 "otel.javaagent.experimental.initializer.jar",
-                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_INITIALIZER_JAR"))));
+                System.getenv("OTEL_JAVAAGENT_EXPERIMENTAL_INITIALIZER_JAR")),
+            javaagentFile));
     // TODO when logging is configured add warning about deprecated property
 
-    if (extension.isEmpty()) {
+    if (extensions.isEmpty()) {
       return parent;
     }
 
-    List<ClassLoader> delegates = new ArrayList<>(extension.size());
-    for (URL url : extension) {
+    List<ClassLoader> delegates = new ArrayList<>(extensions.size());
+    for (URL url : extensions) {
       delegates.add(getDelegate(parent, url));
     }
     return new MultipleParentClassLoader(parent, delegates);
   }
 
-  private static List<URL> parseLocation(String locationName) {
+  private static void includeEmbeddedExtensionsIfFound(
+      ClassLoader parent, List<URL> extensions, File javaagentFile) {
+    try {
+      JarFile jarFile = new JarFile(javaagentFile, false);
+      Enumeration<JarEntry> entryEnumeration = jarFile.entries();
+      String prefix = "extensions/";
+      File tempDirectory = null;
+      while (entryEnumeration.hasMoreElements()) {
+        JarEntry jarEntry = entryEnumeration.nextElement();
+
+        if (jarEntry.getName().startsWith(prefix) && !jarEntry.isDirectory()) {
+          tempDirectory = ensureTempDirectoryExists(tempDirectory);
+
+          File tempFile = new File(tempDirectory, jarEntry.getName().substring(prefix.length()));
+          if (tempFile.createNewFile()) {
+            tempFile.deleteOnExit();
+            extractFile(jarFile, jarEntry, tempFile);
+            addFileUrl(extensions, tempFile);
+          } else {
+            System.out.println("Failed to create temp file " + tempFile);
```

Comment:
Print out to stderr?

```suggestion
            System.err.println("Failed to create temp file " + tempFile);
```